rfcs/015 - Reworked Compiler Front End.md

```diff
@@ -0,0 +1,179 @@
+3. Define the value `abc124` with type `S`. Add a constraint that type `S` must match `string`.
+4. Find all functions and values named `toString`, and use them as the list of possibilities for the value `abc124`.
+5. Going back to the results of step 2, select `none` as defined in `root.ln`, since it is the only function that accepts 0 arguments.
+6. Going back to the results of step 4, select `toString(Maybe<Stringifiable>): string` as defined in `root.ln`, since it is the only function that matches the constraints on `T`.
+7. Inline all of the selected functions, and repeat steps 2-6 as necessary.
+
+Let's look at another example, which the current front-end can't solve without significant backtracking:
+```ln
+from @std/app import print, start, exit
+
+on start {
+  let myMaybe = none();
+  myMaybe = none();
+  myMaybe = some(5);
+  emit exit myMaybe.getOrExit().toInt8();
+}
+```
+
+There's a lot more complexity here!
+The current front-end would have to backtrack across all declarations, assignments, and Microstatement "rerefs" in order to ensure that the type assigned to `myMaybe` is `Maybe<int64>` once the `some(5)` assignment is reached.
+However, with the new front-end, the compiler would initially assign `myMaybe` to type `Maybe<T>`, then add a constraint that `T` can be assigned by the integer literal `5`.
+Then, once the `toInt8` call is reached, the compiler would add another constraint that for whatever type `T` is, there must be a function called `toInt8` that matches the signature `(T) -> int8`.
+If there are no other constraints, the compiler is able to resolve `T` as an `int64` before outputting the corresponding `amm`.
+
+The new compiler front-end will also be able to determine function signatures without requiring type annotations on parameters or the return type.
+A simple example might look like:
+```ln
+fn add4(num) {
+  return add(num, 4);
+}
+```
+
+At the time of writing this RFC, this function fails to parse due to a missing type annotation on `num`, but it's intended to support gradual typing even for function parameters in Alan.
+The current front-end will require substantial work to support this, requiring one of:
+- An interface (internal or in `root.ln`) that ensures there's an `add` operation for the type of `num` (for example, `Addable`).
+- Type-checking can only happen once the function is inlined.
+
+However, the new type resolution system would simply add a constraint that there be a function named `add` that satisfies the signature `(T, 4 as T) -> T` (ignoring the fact that arithmetic operations in Alan actually use `Result` values).
+This wouldn't require the `Addable` interface to be defined at all, and instead would allow for an anonymous interface that's only visible in the compiler to be created.
+The constraints would then not need to be computed for each time the function is called - instead, we must only ensure that the value passed in to `add4` satisfies the so-called "Addable" interface.
```

Comment:
I do have to say, these constraints seem very similar to the [automatic argument interfaces](https://github.com/alantech/alan/blob/main/rfcs/006%20-%20Automatic%20Argument%20Interfaces%20RFC.md) proposed earlier. It would be nice to:

1. Compare/contrast with that already existing RFC. Does this revoke/supersede it?
2. Can interfaces themselves been treated similarly to constraints such that we can have the compiler generate interfaces for the user based on function usage and inject them into source code through a code formatting tool?

I actually haven't seen anything in this RFC that implies that constraints and interfaces are actually different, except one being created by the compiler and the other the developer.

Reply:
I've intentionally kept the user-facing part of the type system the same, to prevent code breakages as we migrate the compiler to the new front-end. It's meant to be compatible with the current intended design of the language, including RFC 6.

Reply:
Ahh, can you elaborate on that? What part of the type system is "submerged"?

Reply:
I'm effectively proposing a middle stage in the compiler front-end that does type resolution and function selection *before* emitting Microstatements in the output. I've already laid out some of the advantages in the RFC, but another notable advantage is that inferring argument types is separate from the inlining stage - with the current system, it'd only be possible for arguments to be type-checked once the function is being inlined, resulting in the compiler having to type-check the same code `n` times for each time the function is called. In the proposed system, the function's type signature need only be generated once, all of its statements need be type-checked once, all of its function selection possibilities need be gathered once, and Microstatement generation need only happen once per function use.

To clarify, this is not a change to the type system of Alan itself, it's a change to the type resolution and inference system in the front end of the compiler. Users shouldn't experience any changes - instead, there will be benefits in other things, such as compilation speed, code maintainability, code modularity, and a familiar algorithm for many. Since this is a new "phase" in the compiler, that means that the Microstatement generation will also have to be rewritten, but it will be much simpler since most of it will already be generated, it'll simply be a matter of inlining and outputting the correct AMM once that happens.